KDD: File Sync (#3102)
msupply-foundation/open-msupply: decisions/2024-02-28_file_sync.md

# File Sync

- _Date_: 2024-02-28
- _Deciders_: James Brunskill, Mark Prins, Andrei Evguenov, Chris Petty
- _Status_: DECIDED (almost)
- _Outcome_: Option 1

## Context

Our cold chain equipment management features require a mechanism to upload and download documents, and images related to assets.

If these were to go via our standard sync process, large files would potentially block the sync process, slow down deployments to sites, etc.

> All options assume we'll use the open-mSupply Central Server Sync process recently established.

Reviewer: Is this documented anywhere?  It sounds like this is designed not to sync to Legacy mSupply sites (but how is that possible)?

This feature has been a long-standing request - see  [Attach document to internal order and sync it to the supplying store requisition #11288](https://github.com/msupply-foundation/msupply/issues/11288)

Author: Thanks Mark,

@andreievg do you have any documentation for the open-mSupply central server sync design.

The quick version is that we plan to run a open-mSupply central server next to our existing msupply central server. open-mSupply will sync to legacy, but also sync to the open-mSupply central for new things such as files to avoid work being done in two places. However this does unfortunately mean this solution would only work for open-mSupply not legacy.

Reviewer: > The quick version is that we plan to run a open-mSupply central server next to our existing msupply central server. open-mSupply will sync to legacy, but also sync to the open-mSupply central for new things such as files to avoid work being done in two places.

OK, so this is what is meant by the 'open-mSupply Central Server Sync' process.  Now I understand, it it makes good sense.

This means that file sync will only work for open-mSupply Installations.

## Options

### Option 1 - Sync File IDs via std process, and download/upload files in a separate process

> Since we already have a process which allows for syncing data, controlling which records go to which sites, etc, we can reuse this process, to send a list of file ID to each site. Once the file ID is know the server can kick off a background process to download the file to the site via HTTP. If a file hasn't been downloaded yet, but is requested by the client, the server could download the file immediately and cache the result. (This could have some obvious challenges with files that are already halfway downloaded, connection issues, etc.)

Reviewer 2: ```suggestion
Since we already have a process which allows for syncing data, controlling which records go to which sites, etc, we can reuse this process, to send a list of file IDs to each site. Once the file ID is known the server can kick off a background process to download the file to the site via HTTP. If a file hasn't been downloaded yet, but is requested by the client, the server could download the file immediately and cache the result. (This could have some obvious challenges with files that are already halfway downloaded, connection issues, etc.)
```

One key advantage of this approach is that we should be able to store partially downloaded (and uploaded) files and continue from where they're left off rather than having the whole sync request fail.

We can also put in additional logic such as suspending file sync while the main sync is in progress, rate limiting, etc.

It is assumed that we won't do anything special to compress files in MVP.

> A similar process can be implemented for uploading files using (TUS)[tus.io]

Reviewer 3: ```suggestion
A similar process can be implemented for uploading files using [TUS](tus.io)
```

> A similar process can be implemented for uploading files using (TUS)[tus.io]

Reviewer 4: Sounds good. What about the case where a file gets updated while a client downloads the previous version? maybe store some checksum with the uploaded file?

> ![File Sync](./media/file_sync.png)

Reviewer 3: Is the actual file_list schema all of the fields shown here for central and remote combined? Remote will need the table and record ID to identify what the record is for.

Author: We'll raise a github issue for the implementation, which can include this kind of detail. Diagram is intended to illustrate the concept but not intended to be a complete database design.

Reviewer 2: to make the columns look neater, I would use `download_percent` and `download_datetime`

though, I don't think percent is very helpful. Surely something more accurate such as chunks or bytes is required in order to know where to pick up from?

_Pros:_

- Keeps the main sync process simple/unchanged
- Allows for targeted logic for file syncing (e.g rate limiting in future)

> - Allows for files to be downloaded where they're needed rather than needing to sync everywhere (via std sync process)

Reviewer 3: I assume we'd make an API for retrieving the documents, which regardless of option in the KDD we may need anyway for the FE getting documents from backend? Be a pro if one API served both purposes.

For people using a how do we call it... a "central server hosted remote site of Open mSupply"? "mSupply as a service"? They just have one instance running cloud only that they access via web. Functionally same as the above, but reiterating it should hopefully work nicely for them downloading files in their browser.

Author: True, we could support that option...

_Cons:_

> - We could endup writing quite a bit of code to implement this and may have unexexpcted challenges with the implementation

Reviewer 3: ```suggestion
- We could end up writing quite a bit of code to implement this and may have unexpected challenges with the implementation
```

- Other sync protocols might provide benefits e.g. rsync for example uses a delta transfer algorithm which only sends the differences between files, rather than the whole file.

> 

Reviewer: - Our Central server will need to have a large and ever-growing storage capacity for all of these documents

Author: Yes, in future we could offload to cloud storage or something? Many of our clients choose to run their own server though, so disk storage capacity could certainly be an issue. More apparent on remote sites though I think. We might need some mechanism to prune off any files that haven't been accessed for a long time, and re-download them if needed (in the event that we start running low on storage space)

> ### Option 2 - Base64 encode files and use std sync

Reviewer 3: I think this one can be veto as it'd block sync or create complexity if it's the current record trying to sync, or if is large enough to need resuming it'll create more complexity in sync? Not to mention addressing that with priority is a whole best in of itself!

Author: Agreed. Mostly included here for completeness. If we knew we only ever had small files, and not many of them it could possibly work I guess

Reviewer 5: I agree, I think this option is a complete non-starter. I get the feeling that as soon as anyone starts syncing files it will all fall over and people won't love it any more.

> Any files needed to be synced could be base64 encoded and sent as part of the normal sync process. This would be a simiplish process, but would mean that sync would be blocked until all files are fully downloaded. Base64 encoding would also increase the size of the file by 33% (or more) which could be a problem for large files.

Reviewer: blocked until all files are fully downloaded, or until the 'current' file is fully downloaded?

Author: In this solution technically files would just be another sync record, so you could get lucky and get you're files last, or could be first in the sync queue, which would mean you'd get nothing until the download completes.
As Chris mentioned above, I don't think this is a viable solution unless we do some kind prioritisation.

> Any files needed to be synced could be base64 encoded and sent as part of the normal sync process. This would be a simiplish process, but would mean that sync would be blocked until all files are fully downloaded. Base64 encoding would also increase the size of the file by 33% (or more) which could be a problem for large files.

Reviewer 2: ```suggestion
Any files needed to be synced could be base64 encoded and sent as part of the normal sync process. This would be a simplish process, but would mean that sync would be blocked until all files are fully downloaded. Base64 encoding would also increase the size of the file by 33% (or more) which could be a problem for large files.
```

This option is probably only viable if we built in some kind of sync record prioritisation, which would make the process more complex.

_Pros:_

- Simple to implement (if we ignore the sync record prioritisation bit)

_Cons:_

- Sync would be blocked until all files are fully downloaded
- Base64 encoding would increase the size of the file by 33% (or more)

### Option 3 - Use a third party file sync process

> A tool such as (https://syncthing.net/)[https://syncthing.net/] or even dropbox, could be setup on every installation to sync files. This would be a simple process, but would mean that we'd need to setup and maintain a third party tool on every installation. This would save us development effort on creating our own file sync process.

Reviewer 3: ```suggestion
A tool such as [https://syncthing.net/](https://syncthing.net/) or even dropbox, could be setup on every installation to sync files. This would be a simple process, but would mean that we'd need to setup and maintain a third party tool on every installation. This would save us development effort on creating our own file sync process.
```

_Pros:_

- Not much to develop, but there's still complexities around what files are available, how do we communicate this to the client?

> - Syncthing is a peer to peer protocol, so concievably we could sync files between installations without needing to go via the central server. (unlikely in practice, but possible)

Reviewer 2: ```suggestion
- Syncthing is a peer to peer protocol, so conceivably we could sync files between installations without needing to go via the central server. (unlikely in practice, but possible)
```

_Cons:_

- How would this work with Android? Can dropbox even be accessed from our code base?
- We'd need to setup and maintain a third party tool on every installation, even if the time taken for this is acceptable, the risk of miss configurations, missed steps in configuration, etc is high.

> - We probably wouldn't have a lot of control over which files go where, which might cause problems with disk usage, espeically on smaller devices.

Reviewer 3: I think this is potentially a big problem especially if future use case is attachments for requisitions, tenders etc.

> - We probably wouldn't have a lot of control over which files go where, which might cause problems with disk usage, espeically on smaller devices.

Reviewer 2: ```suggestion
- We probably wouldn't have a lot of control over which files go where, which might cause problems with disk usage, especially on smaller devices.
```

> 

Author: Yup good point. Will add

> 

Reviewer: Well, this is still an option, it just needs to be approved or configured by the client, right?

> 

Author: Agreed, updated wording to cover this...

### Option 4 - Only store files on central server and query central directly for files when needed

> Would be the simpliest solution, but wouldn't meet our offline first goals

Reviewer 3: Would also still need some like file_list to associate things with specific record of specific table? Could embed in file name/path structure perhaps but potentially doesn't scale well.

Basically saying still want `file_list` regardless, makes it abundantly clear when viewing database schema what's happening! This comment probably applies to option 3 and 5 right?

Author: For #5 I think yes, in theory with #3 we could just include a file ID in the existing records that are sync'd I guess.
If the file's not there the user would just get an error.

> Would be the simpliest solution, but wouldn't meet our offline first goals

Reviewer 2: ```suggestion
Would be the simplest solution, but wouldn't meet our offline first goals
```

### Option 5 - Custom implementation of an existing sync protocol

https://github.com/abusch/st-rust perhaps? It's explicitly marked as a `Toy` implementation though.
If there's a good candidate someone knows of, we could consider this.

> 

Reviewer 3: ```suggestion
### Option 6 - file_list and FTP server

Perhaps use the file_list idea to control relating records, then use [ftp](https://crates.io/crates/ftp) to retrieve from central server? 

Pros: 
- Maybe ftp will be reliable? 
- [ftp](https://crates.io/crates/ftp) seems pretty clean, but is only the client side...
- Presumably heaps of ftp server solutions out there. 

Cons:
- Have to run an ftp server on the central server
  - Complexity? Maintenance?
  - Security? An open port to an ftp server and whatever vulnerabilities it might have...
  - Someone with ftp credentials might be able to download all documents? What can be put in place to constrain this? I'm not familiar with ftp perhaps can configure "users" for sites that are constrained by what they should be able to get via "file_list".  

```

Author: @Chris-Petty Happy to include this, but not sure it's going to add any advantages over HTTP which we already have implemented...
